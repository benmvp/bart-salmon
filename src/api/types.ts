import {UrlParams} from 'url-lib'
import {StationName, Etd} from '../utils/types'

interface EtdsApi {
  type: 'etd';
  command: 'etd';
  params?: UrlParams;
}
interface RoutesApi {
  type: 'route';
  command: 'routes';
  params?: UrlParams;
}
interface DepartureApi {
  type: 'sched';
  command: 'depart';
  params?: UrlParams;
}
interface StationsApi {
  type: 'stn';
  command: 'stns';
  params?: UrlParams;
}
interface StationInfoApi {
  type: 'stn';
  command: 'stninfo';
  params?: UrlParams;
}

export type ApiRequest = EtdsApi
  | RoutesApi
  | DepartureApi
  | StationsApi
  | StationInfoApi

export type ParsedXmlEtdResponse = {
  uri: string;
  date: string;
  time: string;
  station: {
    name: string;
    abbr: StationName;
    etd: Etd[];
  }[]
}
export type ParsedXmlRootResponse = ParsedXmlEtdResponse

