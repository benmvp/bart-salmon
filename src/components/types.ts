export interface SelectorValue<ValueType> {
  value: ValueType;
  display?: string;
}
