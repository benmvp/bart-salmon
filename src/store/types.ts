import {
  StationName,
  SalmonRoute,
  Train,
} from '../utils/types'

export type OptionalStationName = '' | StationName

interface SetStations {
  type: 'SET_STATIONS';
  payload: {
    origin: OptionalStationName;
    destination: OptionalStationName;
  }
}
interface SetNumSalmonRoutes {
  type: 'SET_NUM_SALMON_ROUTES';
  payload: number;
}
interface FetchSalmonInfo {
  type: 'FETCH_SALMON_INFO';
}
interface ReceiveSalmonInfo {
  type: 'RECEIVE_SALMON_INFO';
  payload: {
    routes: SalmonRoute[];
    arrivals: Train[];
    lastUpdated: Date;
  }
}
interface ErrorSalmonInfo {
  type: 'ERROR_SALMON_INFO';
  error: Error;
}
interface SetRiskinessFactor {
  type: 'SET_RISKINESS_FACTOR';
  payload: number;
}
interface SetNumArrivals {
  type: 'SET_NUM_ARRIVALS';
  payload: number;
}

export type AppAction = SetStations
  | SetNumSalmonRoutes
  | FetchSalmonInfo
  | ReceiveSalmonInfo
  | ErrorSalmonInfo
  | SetRiskinessFactor
  | SetNumArrivals
