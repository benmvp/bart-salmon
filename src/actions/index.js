// @flow
import {
    getSuggestedSalmonRoutesFromEtds,
    getNextArrivalsFromEtds
} from '../utils/salmon'
import {getEstimatedTimesOfDeparture} from '../api'
import type {StationName, SalmonRoute, Train} from '../utils/flow'
import type {
    ReduxDispatch,
    ReduxAction,
    ReduxAsyncAction,
    ReduxGetState
} from '../actions/flow'

const NUM_ARRIVALS = 4

export const setNumSalmonRoutes = (numRoutes: number): ReduxAction => ({
    type: 'SET_NUM_SALMON_ROUTES',
    payload: {
        numRoutes
    }
})

export const setRiskinessFactor = (riskinessFactor: number): ReduxAction => ({
    type: 'SET_RISKINESS_FACTOR',
    payload: {
        riskinessFactor
    }
})

const fetchSalmonInfo = (): ReduxAction => ({
    type: 'FETCH_SALMON_INFO'
})

const receiveSalmonInfo = (
    routes: SalmonRoute[],
    arrivals: Train[]
): ReduxAction => ({
    type: 'RECEIVE_SALMON_INFO',
    payload: {
        routes,
        arrivals
    }
})

const errorSalmonInfo = (error: Error): ReduxAction => ({
    type: 'ERROR_SALMON_INFO',
    error
})

export const getSalmonInfo = (): ReduxAsyncAction => async (
    dispatch: ReduxDispatch,
    getState: ReduxGetState
) => {
    let {origin, destination, numSalmonRoutes, riskinessFactor} = getState()

    if (origin && destination) {
        dispatch(fetchSalmonInfo())

        try {
            let etdsLookup = await getEstimatedTimesOfDeparture()
            let salmonRoutes = getSuggestedSalmonRoutesFromEtds(
                etdsLookup,
                origin,
                destination,
                numSalmonRoutes,
                riskinessFactor
            )
            let arrivals = getNextArrivalsFromEtds(
                etdsLookup,
                origin,
                destination,
                NUM_ARRIVALS
            )

            dispatch(receiveSalmonInfo(salmonRoutes, arrivals))
        } catch (error) {
            // TODO: Create a custom Error that wraps this native error message
            // to be more friendly
            dispatch(errorSalmonInfo(error))
        }
    }
}

export const setOrigin = (name: StationName): ReduxAction => async (
    dispatch: ReduxDispatch
) => {
    dispatch({
        type: 'SET_ORIGIN',
        payload: {
            name
        }
    })
    dispatch(getSalmonInfo())
}

export const setDestination = (name: StationName): ReduxAction => async (
    dispatch: ReduxDispatch
) => {
    dispatch({
        type: 'SET_DESTINATION',
        payload: {
            name
        }
    })
    dispatch(getSalmonInfo())
}
