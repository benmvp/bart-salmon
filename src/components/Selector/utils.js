export const findValueInfo = (valueToFind, values) => (
    values.find(({value}) => value === valueToFind)
)

export const validateValue = (valueToValidate, values) => {
    let validatedValue = valueToValidate

    let valueExists = findValueInfo(validatedValue, values)

    if (!valueExists && values.length) {
        validatedValue = values[0].value
    }

    return validatedValue
}
