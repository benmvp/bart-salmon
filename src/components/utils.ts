import {SelectorValue} from './types'

export const findValueInfo = (
  values: SelectorValue[],
  valueToFind?: string, 
): SelectorValue | undefined =>
  values.find(({value}) => value === valueToFind)

export const validateValue = (
  values: SelectorValue[],
  valueToValidate?: string, 
): string | undefined => {
  let validatedValue = valueToValidate

  let valueExists = findValueInfo(values, validatedValue)

  if (!valueExists && values.length) {
    validatedValue = values[0].value
  }

  return validatedValue
}
