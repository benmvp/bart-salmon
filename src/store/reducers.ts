import {combineReducers} from 'redux'
import {SalmonRoute, StationName, Train} from '../utils/types'
import {AppAction} from './types'

const DEFAULT_NUM_SALMON_ROUTES = 4
const DEFAULT_RISKINESS_FACTOR = 0

const origin = (
  state: StationName = 'POWL',
  action: AppAction,
): StationName => {
  let newState = state

  if (action.type === 'SET_ORIGIN') {
    newState = action.payload
  }

  return newState
}

const destination = (
  state: StationName = 'PITT',
  action: AppAction,
): StationName => {
  let newState = state

  if (action.type === 'SET_DESTINATION') {
    newState = action.payload
  }

  return newState
}

const isFetching = (state: boolean = false, action: AppAction): boolean => {
  const {type} = action
  let newState = state

  if (type === 'FETCH_SALMON_INFO') {
    newState = true
  } else if (type === 'RECEIVE_SALMON_INFO' || type === 'ERROR_SALMON_INFO') {
    newState = false
  }

  return newState
}

const salmonRoutes = (
  state: SalmonRoute[] = [],
  action: AppAction,
): SalmonRoute[] => {
  let newState = state

  if (action.type === 'RECEIVE_SALMON_INFO') {
    newState = action.payload.routes
  } else if (action.type === 'ERROR_SALMON_INFO') {
    newState = []
  }

  return newState
}

const arrivals = (
  state: Train[] = [],
  action: AppAction,
): Train[] => {
  let newState = state

  if (action.type === 'RECEIVE_SALMON_INFO') {
    newState = action.payload.arrivals
  } else if (action.type === 'ERROR_SALMON_INFO') {
    newState = []
  }

  return newState
}

const numSalmonRoutes = (
  state: number = DEFAULT_NUM_SALMON_ROUTES,
  action: AppAction,
): number => {
  let newState = state

  if (action.type === 'SET_NUM_SALMON_ROUTES') {
    newState = action.payload
  }

  return newState
}

// the higher the number, the more likely you'll make the returning train after
// going backwards. Negative numbers mean the train probably would've left before
// you get there
const riskinessFactor = (
  state: number = DEFAULT_RISKINESS_FACTOR,
  action: AppAction,
): number => {
  let newState = state

  if (action.type === 'SET_RISKINESS_FACTOR') {
    newState = action.payload
  }

  return newState
}

const rootReducer = combineReducers({
  origin,
  destination,
  isFetching,
  salmonRoutes,
  arrivals,
  numSalmonRoutes,
  riskinessFactor,
})

export type AppState = ReturnType<typeof rootReducer>

export default rootReducer
