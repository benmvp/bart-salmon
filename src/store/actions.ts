import {Action} from 'redux'
import {ThunkAction} from 'redux-thunk'
import {
  getSuggestedSalmonRoutesFromEtds,
  getNextArrivalsFromEtds,
} from '../utils/salmon'
import {getEstimatedTimesOfDeparture} from '../api'
import {StationName, SalmonRoute, Train} from '../utils/types'
import {AppAction} from './types'
import {AppState} from './reducers'

const NUM_ARRIVALS = 4

export const setNumSalmonRoutes = (numRoutes: number): AppAction => ({
  type: 'SET_NUM_SALMON_ROUTES',
  payload: numRoutes,
})

export const setRiskinessFactor = (riskinessFactor: number): AppAction => ({
  type: 'SET_RISKINESS_FACTOR',
  payload: riskinessFactor,
})

const fetchSalmonInfo = (): AppAction => ({
  type: 'FETCH_SALMON_INFO',
})

const receiveSalmonInfo = (
  routes: SalmonRoute[],
  arrivals: Train[],
): AppAction => ({
  type: 'RECEIVE_SALMON_INFO',
  payload: {
    routes,
    arrivals,
  },
})

const errorSalmonInfo = (error: Error): AppAction => ({
  type: 'ERROR_SALMON_INFO',
  error,
})

export const getSalmonInfo = (): ThunkAction<void, AppState, null, Action<string>> => async (
  dispatch,
  getState,
) => {
  let {origin, destination, numSalmonRoutes, riskinessFactor} = getState()

  if (origin && destination) {
    dispatch(fetchSalmonInfo())

    try {
      let etdsLookup = await getEstimatedTimesOfDeparture()
      let salmonRoutes = getSuggestedSalmonRoutesFromEtds(
        etdsLookup,
        origin,
        destination,
        numSalmonRoutes,
        riskinessFactor,
      )
      let arrivals = getNextArrivalsFromEtds(
        etdsLookup,
        origin,
        destination,
        NUM_ARRIVALS,
      )

      dispatch(receiveSalmonInfo(salmonRoutes, arrivals))
    } catch (error) {
      // TODO: Create a custom Error that wraps this native error message
      // to be more friendly
      dispatch(errorSalmonInfo(error))
    }
  }
}

export const setOrigin = (name: StationName): ThunkAction<void, AppState, null, AppAction> => (
  dispatch,
) => {
  dispatch({
    type: 'SET_ORIGIN',
    payload: name,
  })
  dispatch(getSalmonInfo())
}

export const setDestination = (name: StationName): ThunkAction<void, AppState, null, AppAction> => (
  dispatch,
) => {
  dispatch({
    type: 'SET_DESTINATION',
    payload: name,
  })
  dispatch(getSalmonInfo())
}
