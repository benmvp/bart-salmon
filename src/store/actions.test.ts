import { setStations, setNumSalmonRoutes, getSalmonInfo } from './actions'
import { AppState } from './reducers'
import { StationName } from '../utils/types'

jest.mock('../api')

const MOCK_APP_STATE: AppState = ({
  origin: 'DELN',
  destination: 'EMBR',
  numSalmonRoutes: 3,
  riskinessFactor: 1,
  isFetching: false,
  salmonRoutes: [],
  arrivals: [],
})
const MOCK_GET_STATE = () => MOCK_APP_STATE

describe('setStations', () => {
  it('dispatches SET_STATIONS action w/ specified origin/destination', async () => {
    const mockDispatch = jest.fn()
    const origin: StationName = 'WDUB'
    const destination: StationName = 'NCON'
    const asyncAction = setStations({ origin, destination })

    await asyncAction(mockDispatch, MOCK_GET_STATE, null)

    expect(mockDispatch).toHaveBeenCalledTimes(2)
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'SET_STATIONS',
      payload: { origin, destination },
    })
  })

  it('dispatches getSalmonInfo action', () => {
    // TODO
  })
})

describe('setNumSalmonRoutes', () => {
  it('returns SET_NUM_SALMON_ROUTES action w/ specified amount', () => {
    let numRoutes = 19
    let action = setNumSalmonRoutes(numRoutes)

    expect(action).toEqual({
      type: 'SET_NUM_SALMON_ROUTES',
      payload: numRoutes,
    })
  })
})

describe('getSalmonInfo', () => {
  it('dispatches FETCH_SALMON_INFO action', async () => {
    let mockDispatch = jest.fn()
    let asyncAction = getSalmonInfo()

    await asyncAction(mockDispatch, MOCK_GET_STATE, undefined)

    expect(mockDispatch).toHaveBeenCalledTimes(2)
    expect(mockDispatch).toHaveBeenCalledWith({ type: 'FETCH_SALMON_INFO' })
  })

  it('dispatches RECEIVE_SALMON_INFO action', async () => {
    const mockDispatch = jest.fn()
    const asyncAction = getSalmonInfo()
    const getState = () => ({
      origin: 'DELN',
      destination: 'EMBR',
      numSalmonRoutes: 3,
      riskinessFactor: 1,
      isFetching: false,
      salmonRoutes: [],
      arrivals: [],
      numArrivals: 3,
    }) as AppState

    await asyncAction(mockDispatch, getState, undefined)

    expect(mockDispatch).toHaveBeenCalledTimes(2)

    const actualReceiveActionType = mockDispatch.mock.calls[1][0]

    expect(actualReceiveActionType).toEqual(expect.objectContaining({
      type: 'RECEIVE_SALMON_INFO',
      payload: {
        arrivals: expect.any(Array),
        routes: expect.any(Array),
      }
    }))
  })

  it('dispatches ERROR_SALMON_INFO action', async () => {
    let mockDispatch = jest.fn()
    let asyncAction = getSalmonInfo()
    let getState = () => ({
      origin: '24TH',
      destination: '24TH',
      numSalmonRoutes: 7,
      riskinessFactor: 2,
      isFetching: false,
      salmonRoutes: [],
      arrivals: [],
      numArrivals: 3,
    }) as AppState

    await asyncAction(mockDispatch, getState, null)

    expect(mockDispatch).toHaveBeenCalledTimes(2)

    let actualErrorActionType = mockDispatch.mock.calls[1][0]

    expect(actualErrorActionType).toEqual({
      type: 'ERROR_SALMON_INFO',
      error: expect.any(Error),
    })
  })
})
