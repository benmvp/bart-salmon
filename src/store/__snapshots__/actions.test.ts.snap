// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSalmonInfo dispatches RECEIVE_SALMON_INFO action 1`] = `
Object {
  "payload": Object {
    "arrivals": Array [
      Object {
        "abbreviation": "MLBR",
        "bikeflag": 1,
        "color": "RED",
        "destination": "Millbrae",
        "direction": "South",
        "hexcolor": "#ff0000",
        "length": 9,
        "limited": 0,
        "minutes": 6,
        "platform": 2,
      },
      Object {
        "abbreviation": "MLBR",
        "bikeflag": 1,
        "color": "RED",
        "destination": "Millbrae",
        "direction": "South",
        "hexcolor": "#ff0000",
        "length": 9,
        "limited": 0,
        "minutes": 21,
        "platform": 2,
      },
      Object {
        "abbreviation": "MLBR",
        "bikeflag": 1,
        "color": "RED",
        "destination": "Millbrae",
        "direction": "South",
        "hexcolor": "#ff0000",
        "length": 9,
        "limited": 0,
        "minutes": 34,
        "platform": 2,
      },
    ],
    "routes": Array [
      Object {
        "backwardsRideTime": 5,
        "backwardsRouteId": "ROUTE 3",
        "backwardsStation": "RICH",
        "backwardsTrain": Object {
          "abbreviation": "RICH",
          "bikeflag": 1,
          "color": "ORANGE",
          "destination": "Richmond",
          "direction": "North",
          "hexcolor": "#ff9933",
          "length": 6,
          "limited": 0,
          "minutes": 1,
          "platform": 1,
        },
        "backwardsWaitTime": 9,
        "returnRideTime": 4,
        "returnRouteId": "ROUTE 7",
        "returnTrain": Object {
          "abbreviation": "MLBR",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Millbrae",
          "direction": "South",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 15,
          "platform": 2,
        },
        "waitTime": 1,
      },
      Object {
        "backwardsRideTime": 4,
        "backwardsRouteId": "ROUTE 8",
        "backwardsStation": "RICH",
        "backwardsTrain": Object {
          "abbreviation": "RICH",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Richmond",
          "direction": "North",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 9,
          "platform": 1,
        },
        "backwardsWaitTime": 2,
        "returnRideTime": 4,
        "returnRouteId": "ROUTE 7",
        "returnTrain": Object {
          "abbreviation": "MLBR",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Millbrae",
          "direction": "South",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 15,
          "platform": 2,
        },
        "waitTime": 9,
      },
      Object {
        "backwardsRideTime": 5,
        "backwardsRouteId": "ROUTE 3",
        "backwardsStation": "RICH",
        "backwardsTrain": Object {
          "abbreviation": "RICH",
          "bikeflag": 1,
          "color": "ORANGE",
          "destination": "Richmond",
          "direction": "North",
          "hexcolor": "#ff9933",
          "length": 6,
          "limited": 0,
          "minutes": 1,
          "platform": 1,
        },
        "backwardsWaitTime": 23,
        "returnRideTime": 4,
        "returnRouteId": "ROUTE 7",
        "returnTrain": Object {
          "abbreviation": "MLBR",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Millbrae",
          "direction": "South",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 29,
          "platform": 2,
        },
        "waitTime": 1,
      },
    ],
  },
  "type": "RECEIVE_SALMON_INFO",
}
`;
